slider/beatmap.py

```diff
@@ -1414,31 +1415,150 @@ def ar(self,
 
         return ar
 
-    @lazyval
-    def hit_objects_no_spinners(self):
-        """The hit objects with spinners filtered out.
+    @memoize
+    def hit_objects(self,
+                    spinners=True,
+                    stacking=True,
+                    easy=False,
+                    hard_rock=False,
+                    double_time=False,
+                    half_time=False):
+        """Retrieve hit_objects.
+
+        Parameters
+        ----------
+        spinners : bool, optional
+            If spinners should be included.
+        stacking : bool, optional
+            If stacking should be calculated.
+        easy : bool, optional
+            Calculate stacking with the easy mod enabled.
+        hard_rock : bool, optional
+            Hit_objects will be flipped on the y axis. Calculate stacking with the hard rock mod enabled.
+        half_time : bool, optional
+            Calculate stacking with the half time mod enabled.
+        double_time : bool, optional
+            Calculate stacking with the double time mod enabled.
+
+        Returns
+        -------
+        hit_objects : list[HitObject]
+            The effective AR value.
         """
-        return tuple(e for e in self.hit_objects if not isinstance(e, Spinner))
+
+        hit_objects = self._hit_objects
+
+        if hard_rock:
+            for hit_object in hit_objects:
+                p = hit_object.position
+                p_new = Position(p.x, 384 - p.y)
+                hit_object.position = p_new
+                if isinstance(hit_object, Slider):
+                    hit_object.curve = hit_object.curve.hard_rock
+
+        if stacking:
+
+            ar = self.ar(easy=easy, hard_rock=hard_rock, half_time=half_time, double_time=double_time)
+            stack_threshold = timedelta(milliseconds=ar_to_ms(ar) * (self.stack_leniency * 10))
+            stack_distance = 3
+
+            for i in reversed(range(len(hit_objects))):
```

Comment:
where's the talk? I'm interested :(

Reply:
Oh weird, I used a markdown link and it seems like it went away. Thanks for catching this!
https://nedbatchelder.com/text/iter.html

Reply:
well, I learned that "dunder" is a convenient way to avoid saying "underscore underscore" in conversation. And was reminded how cool itertools can be (not that I tend to make much use of it). Thanks for the link.